tools/timemachine/internal/cmd/download.go

```diff
@@ -0,0 +1,196 @@
+				}
+			}(obj)
+		}
+		wg.Wait()
+
+		log.L().Info("download the height's dir successfully.")
+		return nil
+	},
+}
+
+// downloadFile downloads an object to a file.
+func downloadFile(bucket, object, destFileName string) error {
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return errors.Wrap(err, "Failed to storage.NewClient.")
+	}
+	defer client.Close()
+	client.SetRetry(storage.WithPolicy(storage.RetryAlways))
+
+	ctx, cancel := context.WithTimeout(ctx, GcpTimeout)
+	defer cancel()
+
+	if err = mkdirIfNotExist(filepath.Dir(destFileName)); err != nil {
+		return err
+	}
+	f, err := os.Create(destFileName)
+	if err != nil {
+		return errors.Wrap(err, "Failed to os.Create.")
+	}
+	rc, err := client.Bucket(bucket).Object(object).NewReader(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to Object(%q).NewReader.", object)
+	}
+	defer rc.Close()
+
+	if _, err = io.Copy(f, rc); err != nil {
+		return errors.Wrap(err, "Failed to io.Copy.")
+	}
+
+	if err = f.Close(); err != nil {
+		return errors.Wrap(err, "Failed to f.Close.")
+	}
+	return nil
+}
+
+// listFiles lists objects within specified bucket.
+func listFiles(bucket, prefix, delim string) ([]string, error) {
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to storage.NewClient.")
+	}
+	defer client.Close()
+	client.SetRetry(storage.WithPolicy(storage.RetryAlways))
+
+	ctx, cancel := context.WithTimeout(ctx, GcpTimeout)
+	defer cancel()
+
+	query := &storage.Query{
+		Prefix:    prefix,
+		Delimiter: delim,
+	}
+	if err := query.SetAttrSelection([]string{"Name"}); err != nil {
+		return nil, errors.Wrap(err, "Failed to query.SetAttrSelection.")
+	}
+	it := client.Bucket(bucket).Objects(ctx, query)
+	objNames := make([]string, 0)
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to client.Bucket(%q).Objects", bucket)
+		}
+		if attrs.Name != "" && attrs.Name != prefix {
+			objNames = append(objNames, attrs.Name)
+		}
+	}
+	return objNames, nil
+}
+
+func genHeightDir(height uint64) (heightDir string) {
+	if height < 8000000 {
+		heightDir = "0m"
+	} else if height >= 8000000 && height < 12000000 {
+		heightDir = "8m"
+	} else if height >= 12000000 && height < 13000000 {
+		heightDir = "12m"
+	} else {
+		inter := height / 1000000
+		interStr := fmt.Sprintf("%dm", inter)
+		deci := height - inter*1000000
+		if deci < 250000 {
+			heightDir = interStr
+		} else if deci >= 250000 && deci < 500000 {
+			heightDir = interStr + "25"
+		} else if deci >= 500000 && deci < 750000 {
+			heightDir = interStr + "50"
+		} else {
+			heightDir = interStr + "75"
+		}
+	}
+	return
+}
+
+func mkdirIfNotExist(destDir string) error {
+	if _, err := os.Stat(destDir); os.IsNotExist(err) {
+		if err = os.MkdirAll(destDir, 0744); err != nil {
```

Comment:
we should not create dir? if the path does not exist, something is wrong?

Reply:
mkdir `fullsync/mainnet/13m/`   to  keep with dir in gcp

Reply:
mkdir `fullsync/mainnet/13m/`   to  keep with dir in gcp

Reply:
timemachine/13m

Reply:
fixed